ninja/pagination.py

```diff
@@ -109,37 +152,53 @@ def my_view(request):
 
     """
 
-    isfunction = inspect.isfunction(func_or_pgn_class)
+    def _is_view_func(func: Any) -> TypeGuard[ViewFn[Req, P]]:
+        return inspect.isfunction(func_or_pgn_class)
+
     isnotset = func_or_pgn_class == NOT_SET
 
     pagination_class: Type[PaginationBase] = import_string(settings.PAGINATION_CLASS)
 
-    if isfunction:
+    if _is_view_func(func_or_pgn_class):
         return _inject_pagination(func_or_pgn_class, pagination_class)
 
-    if not isnotset:
+    # Second check is redundant, but `TypeGuard` doesn't narrow the negative case.
+    # `TypeIs` should resolve this: https://peps.python.org/pep-0742/
+    if not isnotset and isinstance(func_or_pgn_class, type):
         pagination_class = func_or_pgn_class
 
-    def wrapper(func: Callable) -> Any:
+    def wrapper(func: ViewFn[Req, P]) -> PaginatedViewFn[Req, P]:
         return _inject_pagination(func, pagination_class, **paginator_params)
 
     return wrapper
 
 
 def _inject_pagination(
-    func: Callable,
+    func: ViewFn[Req, P],
     paginator_class: Type[PaginationBase],
     **paginator_params: Any,
-) -> Callable:
-    paginator: PaginationBase = paginator_class(**paginator_params)
+) -> PaginatedViewFn[Req, P]:
+    """Inject pagination into the view function.
+
+    Args:
+        func: The view function.
+        paginator_class: The paginator class.
+        **paginator_params: Parameters for the paginator class.
+
+    Returns:
+        The view function with pagination injected into the response.
+    """
+    paginator = paginator_class(**paginator_params)
 
     @wraps(func)
-    def view_with_pagination(request: HttpRequest, **kwargs: Any) -> Any:
+    def view_with_pagination(
+        request: Req, *args: P.args, **kwargs: P.kwargs
+    ) -> Dict[str, Any]:
         pagination_params = kwargs.pop("ninja_pagination")
         if paginator.pass_parameter:
             kwargs[paginator.pass_parameter] = pagination_params
 
-        items = func(request, **kwargs)
+        items = func(request, *args, **kwargs)
```

Comment:
To properly utilize `ParamSpec` (which lets us preserve the call signature), we need to pass in `*args`. I'm guessing this wouldn't be problematic, but if will be, I can rethink the approach here.